Sources/Utility/FSWatch.swift

```diff
@@ -0,0 +1,602 @@
+
+    /// The paths being watched.
+    public let paths: [AbsolutePath: WatchOptions]
+
+    /// The delegate.
+    private let delegate: InotifyDelegate?
+
+    /// The settle period (in seconds).
+    public let settle: Int = 2
+
+    /// Internal properties.
+    private var fd: Int32?
+
+    /// The list of watched directories/files.
+    private var wds: [Int32: AbsolutePath] = [:]
+
+    /// The queue on which we read the events.
+    private let readQueue = DispatchQueue(label: "org.swift.swiftpm.\(Inotify.self).read)")
+
+    /// Callback queue for the delegate.
+    private let callbacksQueue = DispatchQueue(label: "org.swift.swiftpm.\(Inotify.self).callback)")
+
+    /// Condition for handling event reporting.
+    private var reportCondition = Condition()
+
+    // Should be read or written to using the report condition only.
+    private var collectedEvents: [AbsolutePath] = []
+
+    // Should be read or written to using the report condition only.
+    private var lastEventTime: Date? = nil
+
+    // Should be read or written to using the report condition only.
+    private var cancelled = false
+
+    /// Pipe for waking up the read loop.
+    private var cancellationPipe: [Int32] = [0, 0]
+
+    /// Create a inotify instance.
+    ///
+    /// The paths are not watched recursively.
+    public init(paths: [AbsolutePath: WatchOptions], delegate: InotifyDelegate? = nil) {
+        self.paths = paths
+        self.delegate = delegate
+    }
+
+    /// Start the watch operation.
+    public func start() throws {
+
+        // All paths need to exist.
+        for (path, _) in paths {
+            guard localFileSystem.exists(path) else {
+                throw Error.failedToWatch(path)
+            }
+        }
+    
+        // Create the file descriptor.
+        let fd = inotify_init1(Int32(IN_NONBLOCK))
+
+        guard fd != -1 else {
+            throw Error.invalidFD
+        }
+        self.fd = fd
+    
+        /// Add watch for each path.
+        for (path, options) in paths {
+
+            let wd = inotify_add_watch(fd, path.asString, UInt32(options.rawValue))
+            guard wd != -1 else {
+                throw Error.failedToWatch(path)
+            }
+
+            self.wds[wd] = path
+        }
+
+        // Start the report thread.
+        startReportThread()
+
+        readQueue.async {
+            self.startRead()
+        }
+    }
+
+    /// End the watch operation.
+    public func stop() {
+        // FIXME: Write precondition to ensure this is called only once.
+        guard let fd = fd else {
+            assertionFailure("end called without a fd")
+            return
+        }
+
+        // Shutdown the report thread.
+        reportCondition.whileLocked {
+            cancelled = true
+            reportCondition.signal()
+        }
+
+        // Wakeup the read loop by writing on the cancellation pipe.
+        let writtenData = write(cancellationPipe[1], "", 1)
```

Comment:
I don't think this is necessary, if you just close the file descriptor, the waiting `read` call will return `-1`, you could just check for that at the read loop.

Reply:
Unfortunately, this is the only way to ensure safety. From http://man7.org/linux/man-pages/man2/close.2.html :

```
 It is probably unwise to close file descriptors while they may be in
use by system calls in other threads in the same process.  Since a
file descriptor may be reused, there are some obscure race conditions
that may cause unintended side effects.
```

Reply:
Ah interesting, good to know.